File: Leetcode/merge_k_sorted_lists/merge_k_sorted_lists.cpp
#include <vector>
#include <algorithm>
using namespace std;
/**
 * Definition for singly-linked list.
 */ 
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        vector<ListNode*> allNodes;
        for (int i = 0; i < lists.size(); i++) {
            ListNode* current = lists[i];
            while (current) {
                allNodes.push_back(current);
                current = current->next;
            }
        }
        sort(allNodes.begin(), allNodes.end(), 
        [](ListNode* a, ListNode* b) {
           return a->val > b->val;
        });

        ListNode* current = allNodes[0];
        for (int i = 0; i < allNodes.size()-1; i++) {
            current->next = allNodes[i+1];
            current = current->next;
        }
        current->next = nullptr;
        return allNodes[0];
    }
};

File: Leetcode/merge_k_sorted_lists/merge_k_sorted_lists.py
# Definition for singly-linked list.
import heapq

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: list[ListNode]) -> ListNode:
        setattr(ListNode, "__lt__", lambda self, other: self.val <= other.val)
        heapq.heapify(lists)        

        dummy = ListNode()
        current = dummy

        while (lists):
            smallest = heapq.heappop(lists)
            current.next = smallest
            current = current.next

            if (smallest.next):
                heapq.heappush(lists, smallest.next)
        return dummy.next
    
if __name__ == "__main__":
    s = Solution()
    s.mergeKLists([ListNode(2),ListNode(2),ListNode(6),ListNode(2),ListNode(4),ListNode(5)])